ateam_ai/src/behavior/behavior_evaluator.cpp

```diff
@@ -0,0 +1,119 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "behavior/behavior_evaluator.hpp"
+
+#include "behavior/behavior_feedback.hpp"
+
+BehaviorEvaluator::BehaviorEvaluator(BehaviorRealization & behavior_realization)
+: behavior_realization(behavior_realization) {}
+
+void BehaviorEvaluator::update_world(World world)
+{
+  this->world = world;
+}
+
+DirectedGraph<Behavior> BehaviorEvaluator::get_best_behaviors()
+{
+  //
+  // Do preprocessing on world state to get important metrics like possession
+  //
+
+  //
+  // Setup different behavior options
+  //
+
+  DirectedGraph<Behavior> three_one_touch_shot;
+  Behavior initial_pass_start{
+    Behavior::Type::MovingKick,
+    Behavior::Priority::Required,
+    KickParam({0, 0})};
+  Behavior first_receiver{
+    Behavior::Type::OneTouchReceiveKick,
+    Behavior::Priority::Required,
+    ReceiveParam({0, 0}, {10, 10})};
+  Behavior second_receiver{
+    Behavior::Type::OneTouchReceiveKick,
+    Behavior::Priority::Required,
+    ReceiveParam({10, 10}, {-10, -10})};
+  Behavior final_receiver_shot{
+    Behavior::Type::OneTouchShot,
+    Behavior::Priority::Required,
+    ReceiveShotParam({-10, -10})};
+
+  std::size_t parent = three_one_touch_shot.add_node(initial_pass_start);
+  parent = three_one_touch_shot.add_node(first_receiver, parent);
+  parent = three_one_touch_shot.add_node(second_receiver, parent);
+  parent = three_one_touch_shot.add_node(final_receiver_shot, parent);
+
+
+  DirectedGraph<Behavior> direct_shot;
+  Behavior shot{
+    Behavior::Type::Shot,
+    Behavior::Priority::Required,
+    ShotParam()};
+
+  parent = direct_shot.add_node(shot, parent);
+
+  //
+  // See how that combination of behaviors would be planned and executed
+  //
+  DirectedGraph<BehaviorFeedback> three_one_touch_shot_feedback =
+    behavior_realization.realize_behaviors(three_one_touch_shot);
+  DirectedGraph<BehaviorFeedback> direct_shot_feedback =
+    behavior_realization.realize_behaviors(direct_shot);
+
+  //
+  // Choose main behavior
+  //
+
+  // choose direct shot because score chance is better or
+  // maybe the total behavior completetion time is short
+  // or maybe the other one can't be completed due to number of robots
+  DirectedGraph<Behavior> behavior_out = direct_shot;
+
+  //
+  // Add background behaviors
+  //
+
+  Behavior goalie{
+    Behavior::Type::MoveToPoint,
+    Behavior::Priority::Medium,
+    MoveParam({0, 0})};  // Field::OurGoal.center();
+  Behavior forward{
+    Behavior::Type::MoveToPoint,
+    Behavior::Priority::Low,
+    MoveParam({10, 0})};  // Field::TheirGoal.center();
+  Behavior left_defender{
+    Behavior::Type::MoveToPoint,
+    Behavior::Priority::Medium,
+    MoveParam({5, 5})};
+  Behavior right_defender{
+    Behavior::Type::MoveToPoint,
+    Behavior::Priority::Medium,
+    MoveParam({5, -5})};
```

Comment:
I think this brings up the question of whether these should be included in the realize_behaviors call. IMO they should, but that obviously would duplicate the respective planning calls.

Maybe it makes sense to have a "coarse" and "fine" level of behavior generation? The former could just bubble up some course stat estimates while the latter does full planning?

Reply:
I like the idea of a coarse and fine version. On the high level side, I don't really care about how the behavior happens, but the results of it.

I was splitting these behaviors out from the above set because I was kinda thinking about it in the "primary"/"passive" role split. I really only care about the main ball interaction behaviors (Eg: on offense the capture/pass/shot). The background defenders don't change my decision making in most cases so it would be a waste to calculate them. If they do matter, we can add them to `realize_behaviors` call and get the feedback we need.

I can add them to the original call for now so we don't have any extra requirements on the role assignment piece. It might be something to think about in the future since it encodes background "formations" (for lack of a better word) in a more obvious way. We won't have to plan out all 11 robot's behaviors when trying do decide which plan to do.

Reply:
Maybe we could have an explicit `PassiveFormation` behavior or something? IMO we should guarantee that all behaviors are included in the evaluation, even if we aren't coming up with full plans for all of them.